### DISCUSSION TITLE
New PMD XPath rule

### DISCUSSION THREAD

**Discussion by allurisriakhil** [pmd/pmd#4601] [Q&A] [answered]
Hi,

I am new to use and work with PMD rules. I am trying to create a new PMD XPath rule using xml to identify weather a file name in a directory starts with 'SA_' as naming convention or not. Below is the code snippets that i used in my ruleset. When trying to execute the rule it is not doing the work appropriately i tried debugging but unable to find the issue. Could you please let me know what is the mistake here. Thanks.

```
         <properties>
          <property name="xpath">
            <value>
              <![CDATA[ //file[not(starts-with(@name, 'SA_'))] ]]>
            </value>
          </property>
        </properties>
        <example>
        <![CDATA[
          <file name="SA_MyClass.cls">...</file> <!-- this is correct -->
          <file name="MyClass.cls">...</file> <!-- this is incorrect -->
        ]]>
       </example>

```

**Comment by allurisriakhil** [comment]
@jsotuyod could you please help me with the above query. Thanks.

**Comment by oowekyala** [comment]
I assume you are using `XPathRule` but I can't know for sure.

To write XPath rules for XML, you should use the class `DomXPathRule` as explained [on the website](https://pmd.github.io/pmd/pmd_languages_xml.html#xpath-rules-in-xml):
```xml
<rule name="MyXPathRule"
      language="xml"
      message="A message"
      class="net.sourceforge.pmd.lang.xml.rule.DomXPathRule">

      <properties>
        <property name="xpath">
            <value><![CDATA[
               //file[not(starts-with(@name, 'SA_'))]
            ]]></value>
        </property>
      </properties>
</rule>
```

If that doesn't work please provide more details, including the PMD version you're using, the full `rule` XML element and a sample file you're trying to match.

**Reply by allurisriakhil** [reply]
Hi @oowekyala , Thanks for your response.

I am trying to write a pmd apex rule using XPath to integrate with Salesforce code Analyzer (sfdx-scanner). At the moment the sfdx-scanner tool is still on PMD6.

Below is the Scenario:

Scenario: I am trying to implement a custom rule to check the FileNamingConvention

Example: The FileNamingConvention should start with SA_

Code I am trying to implement:

```

<?xml version="1.0" encoding="UTF-8"?>
<ruleset name="SACodeStyle"
         xmlns="http://pmd.sourceforge.net/ruleset/2.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://pmd.sourceforge.net/ruleset/2.0.0 https://pmd.sourceforge.io/ruleset_2_0_0.xsd">
    <description>Custom ruleset for checking file naming convention starting with SA_</description>
    <rule name="File name should start with SA_ " class="net.sourceforge.pmd.lang.apex.rule.ApexXPathRule" language="apex" message="File name should start with SA_">
        <description>File name should start with SA_</description>
        <priority>3</priority>
        <properties>
          <property name="version" value="2.0"/>
          <property name="xpath">
            <value>
              <![CDATA[ //file[not(starts-with(@name, 'SA_'))] ]]>
            </value>
          </property>
        </properties>
        <example>
        <![CDATA[
          <file name="SA_MyClass.cls">...</file> <!-- this is correct -->
          <file name="MyClass.cls">...</file> <!-- this is incorrect -->
        ]]>
       </example>
    </rule>
</ruleset>

```

Test Case: I am running this ruleset against a directory “ main/default/classes/ ” for example 

So in the classes directory we have multiple files.cls so we want to verify if File name starts with  SA_ or not. 

```
        <![CDATA[
          <file name="SA_MyClass.cls">...</file> <!-- this is correct -->
          <file name="MyClass.cls">...</file> <!-- this is incorrect -->
        ]]>
       </example>
```

Please let me know if you require any further details thanks.

**Reply by oowekyala** [reply, marked as answer]
Ah so this is an Apex rule. Your example is confusing as it contains XML markup.
Anyway we have a `fileName()` function to get the current file name. Try changing your XPath rule to `/*[starts-with(pmd:fileName(), 'SA_')]`.

**Reply by allurisriakhil** [reply]
@oowekyala, Thanks for your time and response on this ask. It worked.
